src/virtual_machine/lc3_virtual_machine.rs

```diff
@@ -0,0 +1,384 @@
+            }
+            Instruction::STR => {
+                let source_register = Register::from((instruction >> 9) & 0b111);
+                let base_register = Register::from((instruction >> 6) & 0b111);
+                let offset = Self::sign_extend(instruction & 0b111111, 6);
+                store_base_offset(self, source_register, base_register, offset)
+            }
+            Instruction::NOT => {
+                let destination_register = Register::from((instruction >> 9) & 0b111);
+                let source_register = Register::from((instruction >> 6) & 0b111);
+                not(self, destination_register, source_register)
+            }
+            Instruction::LDI => {
+                let destination_register = Register::from((instruction >> 9) & 0b111);
+                let program_counter_offset = Self::sign_extend(instruction & 0b111111111, 9);
+                load_indirect(self, destination_register, program_counter_offset)
+            }
+            Instruction::STI => {
+                let source_register = Register::from((instruction >> 9) & 0b111);
+                let program_counter_offset = Self::sign_extend(instruction & 0b111111111, 9);
+                store_indirect(self, source_register, program_counter_offset)
+            }
+            Instruction::JMP => {
+                let base_register = Register::from((instruction >> 6) & 0b111);
+                jump(self, base_register)
+            }
+            Instruction::LEA => {
+                let destination_register = Register::from((instruction >> 9) & 0b111);
+                let program_counter_offset = Self::sign_extend(instruction & 0b111111111, 9);
+                load_effective_address(self, destination_register, program_counter_offset)
+            }
+            Instruction::TRAP => {
+                let trap = Trap::from(instruction & 0b11111111);
+                trap_instruction(self, trap)
+            }
+            Instruction::RTI => panic!("This opcode is not supported"),
+            Instruction::RES => panic!("This opcode is not supported"),
+        }
+    }
+
+    fn sign_extend(mut value_to_extend: u16, ammount_of_bits: u16) -> u16 {
+        if (value_to_extend >> (ammount_of_bits - 1) & 0b1) == 1 {
+            value_to_extend |= 0xFFFF << ammount_of_bits;
+        }
+        value_to_extend
+    }
+}
+
+#[cfg(test)]
+pub mod test {
+    use crate::virtual_machine::lc3_virtual_machine::Flag;
+
+    use super::LC3VirtualMachine;
+
+    #[test]
+    fn can_add_two_numbers_in_same_register() {
+        let mut virtual_machine = LC3VirtualMachine::new(0);
+        let add_one_to_register_zero = 0b0001_000_000_1_00001;
+        virtual_machine.process_input(add_one_to_register_zero);
+        let result = virtual_machine.read_register(super::Register::R0);
+
+        assert_eq!(result, 1);
+    }
+
+    #[test]
+    fn can_add_two_numbers_in_differents_registers() {
```

Comment:
```suggestion
    fn can_add_two_numbers_in_different_registers() {
```
A typo, same happens with some the following tests.